compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerToMicroKernels.cpp

```diff
@@ -0,0 +1,185 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/builtins/ukernel/exported_bits.h"
+#include "iree/compiler/Codegen/Dialect/IREECodegenDialect.h"
+#include "iree/compiler/Codegen/Dialect/MicroKernelOps.h"
+#include "iree/compiler/Codegen/PassDetail.h"
+#include "iree/compiler/Codegen/Passes.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+namespace {
+struct LLVMCPULowerToMicroKernelsPass
+    : LLVMCPULowerToMicroKernelsBase<LLVMCPULowerToMicroKernelsPass> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<IREE::Codegen::IREECodegenDialect>();
+  }
+  void runOnOperation() override;
+};
+}  // namespace
+
+/// Returns `true` if an `outsOperand` value is initialized to zero.
+static bool isInitializedToZero(Value outsOperand) {
+  auto fillOp = outsOperand.getDefiningOp<linalg::FillOp>();
+  if (!fillOp) return false;
+  Value fillVal = fillOp.getDpsInputOperand(0)->get();
+  return matchPattern(fillVal, m_Zero()) ||
+         matchPattern(fillVal, m_AnyZeroFloat());
+}
+
+/// Matches an (linalg.fill -> )? linalg.mmt4d operation sequence and converts
+/// it into a iree_codegen.ukernel.mmt4d operation, that is later lowered
+/// into a call to the microkernel.
+static FailureOr<IREE::Codegen::MicroKernelOpInterface> matchDAGForMicroKernel(
+    RewriterBase &rewriter, linalg::Mmt4DOp op) {
+  if (!op.hasTensorSemantics()) {
+    return rewriter.notifyMatchFailure(
+        op, "operations needs to have tensor semantics");
+  }
+  Value lhs = op.getDpsInputOperand(0)->get();
+  Value rhs = op.getDpsInputOperand(1)->get();
+  Value out = op.getDpsInitOperand(0)->get();
+  auto lhsType = lhs.getType().cast<ShapedType>();
+  auto rhsType = rhs.getType().cast<ShapedType>();
+  auto outType = out.getType().cast<ShapedType>();
+  Type lhsElemType = lhsType.getElementType();
+  Type rhsElemType = rhsType.getElementType();
+  Type outElemType = outType.getElementType();
+  if (!((lhsElemType.isSignlessInteger(8) && rhsElemType.isSignlessInteger(8) &&
+         outElemType.isSignlessInteger(32)) ||
+        (lhsElemType.isF32() && rhsElemType.isF32() && outElemType.isF32()))) {
+    return rewriter.notifyMatchFailure(
+        op, "unsupported type for lowering to microkernels");
+  }
```

Comment:
Can we reuse MatmulType enum class and have better logic/readability here? Something like `Optional<MatmulType> getMatmulType(Type lhsElemType, Type, rhsElemType, Type resElemType`, and we can check if the return value is `F32F32F32` or `I8I8I32` here.

https://github.com/iree-org/iree/blob/584085f1092d01e36c6facd8bfaf21480bcf3474/compiler/src/iree/compiler/Codegen/Common/EncodingInfo.h#L16-L19

Reply:
Great points @hanhanW! Can I suggest that I do this as part of follow-ups?  I am familiar with this and I wanted to pick this up from the earliest possible point to free @MaheshRavishankar for all the other important things that he needs to work on :-)

Reply:
sure thing! I think we can merge the commit now, since Mahesh enabled auto-merge and we all agree with merging it.